aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/ApplicationLifeCycleHelper.java

```diff
@@ -0,0 +1,161 @@
+        Log.d(LOG_TAG, "onActivityCreated " + activity.getLocalClassName());
+        handleOnCreateOrOnStartToHandleApplicationEnteredForeground();
+        activityLifecycleStateMap.put(activity, "created");
+    }
+
+    @Override
+    public void onActivityStarted(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityStarted " + activity.getLocalClassName());
+        handleOnCreateOrOnStartToHandleApplicationEnteredForeground();
+        activityLifecycleStateMap.put(activity, "started");
+    }
+
+    @Override
+    public void onActivityResumed(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityResumed " + activity.getLocalClassName());
+        activityLifecycleStateMap.put(activity, "resumed");
+    }
+
+    @Override
+    public void onActivityPaused(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityPaused " + activity.getLocalClassName());
+        activityLifecycleStateMap.put(activity, "paused");
+    }
+
+    @Override
+    public void onActivityStopped(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityStopped " + activity.getLocalClassName());
+        // When the activity is stopped, we remove it from the lifecycle state map since we
+        // no longer consider it keeping a session alive.
+        activityLifecycleStateMap.remove(activity);
+    }
+
+    @Override
+    public void onActivitySaveInstanceState(final Activity activity, final Bundle outState) {
+        Log.d(LOG_TAG, "onActivitySaveInstanceState " + activity.getLocalClassName());
+    }
+
+    @Override
+    public void onActivityDestroyed(final Activity activity) {
+        Log.d(LOG_TAG, "onActivityDestroyed " + activity.getLocalClassName());
+        // Activity should not be in the activityLifecycleStateMap any longer.
+        if (activityLifecycleStateMap.containsKey(activity)) {
+            Log.wtf(LOG_TAG, "Destroyed activity present in activityLifecycleMap!?");
+            activityLifecycleStateMap.remove(activity);
+        }
+    }
+
+    /**
+     * Call this method when your Application trims memory.
+     * @param level the level passed through from Application.onTrimMemory().
+     */
+    public void handleOnTrimMemory(final int level) {
+        Log.d(LOG_TAG, "onTrimMemory " + level);
+        // If no activities are running and the app has gone into the background.
+        if (level >= Application.TRIM_MEMORY_UI_HIDDEN) {
+            checkForApplicationEnteredBackground();
+        }
+    }
+
+    class ScreenOffReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            checkForApplicationEnteredBackground();
+        }
+    }
+
+    /**
+     * Called back when your application enters the Foreground.
+     */
+    protected abstract void applicationEnteredForeground();
+
+    /**
+     * Called back when your application enters the Background.
+     */
+    protected abstract void applicationEnteredBackground();
```

Comment:
Instead of doing this as an `abstract class`, and requiring someone to extend it, I suggest to use the "observer" design pattern. If you do that, this class would become a `final class`, and would provide a way to be "observed."

A classic Java implementation of this would be like:
```
final class ApplicationForegroundTracker {
    private final Set<Observer> observers;

    ApplicationForegroundTracker() {
        this.observers = new HashSet<>();
    }

    void registerForegroundObserver(Observer observer) {
        synchronized (observers) {
            this.observers.add(ObjectsCompat.requireNonNull(observer));
        }
    }

    void unregisterForegroundObserver(Observer observer) {
        synchronized (observers) {
            observers.remove(Objects.requireNonNull(observer));
        }
    }

    // Somewhere in your code, when its time to notify that you have a new state ...
    // You can call this either like notify(observer::onApplicationEnteredForeground);
    // or like notify(observer::onApplicationEnteredBackground);
    void notify(NotificationStrategy strategy) {
        Set<Observer> threadSafeObserversCopy = new HashSet<>();
        synchronized (observers) {
            threadSafeObserversCopy.addAll(observers);
        }

        for (Observer observer: threadSafeObserversCopy) {
            strategy.apply(observer);
        }      
    }

    interface Observer {
        void onApplicationEnteredForeground();
        void onApplciationEnteredBackground();
    }

    // An internal interface to define a function which calls back an observer.
    // This is a strategy design pattern. One strategy might be to call 
    // {@link Observer#onApplicationEnteredForeground()}. Another strategy might be
    // to call {@link Observer#onAppliactionEnteredBackground()}.
    // This strategy is used to execute a small bit of code inside of the notification routine,
    // where a lot of the same setup is used regardless of the notification strategy.
    interface NotificationStrategy {
        void apply(Observer observer);
    }
}
```

Anyway, then you'd use it like:
```
final ApplicationForegroundTracker tracker = new AppliactionForegroundTracker();
final ApplicationForegroundTracker.Observer observer = new ApplicationForegroundTracker.Observer() {
    @Override public void onApplicationEnteredForeground() {
        analyticsSession.resume(); // or whatever
    }
    @Overridd public void onApplicactionEnteredBackground() {
        analyticsSession.pause(); // or whatever
    }
};

tracker.addForegroundObserver(observer);
...
tracker.removeForegroundObserver(observer);
```

Reply:
Made it a final class. We want to provide this feature automatically where session are recorded for every time app goes to back ground after it is brought to foreground. Enduser should not have to write any code for it. Will discuss offline